### DISCUSSION TITLE
API improvement to accommodate for RPM CoW (PR#1470)

### DISCUSSION THREAD

**Discussion by chantra** [rpm-software-management/rpm#2057] [Ideas]
Related PR: https://github.com/rpm-software-management/rpm/pull/1470

> This is a follow-up of https://github.com/rpm-software-management/rpm/pull/1470#issuecomment-1118480244 .
Below, I have tried to capture the required building blocks to be able to integrate RPM CoW into RPM with limited conditionals in the core RPM codebase. While I suggest some specific plugin hooks, this is not meant to be a prescriptive list and I am open for feedbacks/suggestions to reach the end goal.

[RPM CoW](https://fedoraproject.org/wiki/Changes/RPMCoW) minimizes IO by taking advantages of reflinking available in file systems such as BTRFS and XFS.
In order to do this, the data on disk must be layed down as [extents](https://en.wikipedia.org/wiki/Extent_(file_systems)) which can then be [reflink-ed](https://blogs.oracle.com/linux/post/xfs-data-block-sharing-reflink).

To do this, RPM CoW introduces `rpm2extents`, an RPM utility to take an RPM as input stream and output that same RPM where the payload has been extracted and aligned to form file extents. Some meta data is also added to the end of the file to be able to find at which offset in the transcoded RPM file a given file in the payload is. When installing the file, rpm can then reflink this file instead of copying it. Therefore, the RPM that is on-disk, while having the same content, is not bitwise copy.
Under a typical case, it saves at least 1 IO (1 read when reading the compressed RPM, 1 write when copying files to disk). This becomes more valuable when servers are running container workload, or CI workload (think mock builders) as the same package are installed multiple times from the same cached file. Data is now reflinked (O(1)) instead of being decompressed and copied, saving both IOs and disk space.

This also means that RPM needs to be taught how to install such files. Fortunately, [RPM supports plugins](https://rpm-software-management.github.io/rpm/manual/plugins.html) which allows augmenting RPM in an optional way and without modifying the core RPM logic. The current API has limitations for the RPM CoW use case though, but we should be able to provide new API entry points to make this work.

[RPM CoW changes](https://github.com/chantra/rpm/compare/6058b070b426a5c99a0bbc24c295824ab5d0641c...cow_signvalidation) currently needs a few more hooks to be able to perform its job:
- As part of rpmPackageFilesInstall it needs to provide a [custom rpmfi iterator](https://github.com/chantra/rpm/compare/6058b070b426a5c99a0bbc24c295824ab5d0641c...5f762af17c6e72e86e4710975dcdfd71fc5d1b07#diff-53bfda548820c5dd70e28669eaa5f7bc36c55d3ac25dbd3f590c85ffa29bbbe0): `rpmfilesIter(files, RPMFI_ITER_FWD)` instead of uses the content of the archive with `rpmfiNewArchiveReader(payload, files, RPMFI_ITER_READ_ARCHIVE)` as the list of files/directories and their metadata is now read from the RPM headers metadata vs the cpio archive one. In the current RPM CoW implementation, this is what is being done with [plugin_fsm_file_archive_reader_func](https://github.com/chantra/rpm/blob/ec9c80b2faaa9124e28913c41311f04a513887b8/lib/rpmplugin.h#L66-L68)
- When a file needs to be installed, it needs be given the opportunity to install the file itself (e.g reflink instead of copying from archive) and bypass the original RPM logic. In current RPM CoW implementation it is done by [plugin_fsm_file_install_func](https://github.com/chantra/rpm/blob/ec9c80b2faaa9124e28913c41311f04a513887b8/lib/rpmplugin.h#L63-L65)
- The plugin needs a mechanism to find the file descriptor of the RPM. Currently this is done in exposing a [rpmteFd function](https://github.com/chantra/rpm/blob/ec9c80b2faaa9124e28913c41311f04a513887b8/plugins/reflink.c#L144) that given a transaction element returns the fd associated with it.

Another new concept that needs to be introduced to the RPM plugin API is the ability for a plugin hook to stop the plugin chain from being executed. Currently, every plugin that registers to a specific hook is called as part of the plugin chain, regardless of whatever action may have been performed by a previous plugin. In cases such as installing file or creating the rpmfi iterator, as soon as a plugin has performed that action, the chain must be interrupted. In the current code, this is done by the introduction of the `RPMRC_PLUGIN_CONTENTS` return code.

As mentioned previously, the final RPM which is on-disk after having being transcoded by `rpm2extents` is not a bitwise copy of the original RPM, as such signature/digest verifications on anything that includes the payload will not work.
The current approach is that `rpm2extents` performs the signature validation while streaming the original RPM. The result of this verification (return code + text output) is written into the transcoded RPM trailing metadata so it can be re-used later by `rpmkeys`, or anything that needs to validate the package like `rpm -i`. Having a plugin hook that can defer the validation of RPM to the plugin would help in separating that logic from RPM core.

**Comment by chantra** [comment]
cc @pmatilai @ffesti @malmond77 @davide125 @Conan-Kudo

**Comment by DemiMarie** [comment]
> As mentioned previously, the final RPM which is on-disk after having being transcoded by `rpm2extents` is not a bitwise copy of the original RPM, as such signature/digest verifications on anything that includes the payload will not work.

A better approach would be to change the RPM format to provide a digest that can be verified in a streaming manner.

**Comment by chantra** [comment]
> > As mentioned previously, the final RPM which is on-disk after having
> > being transcoded by rpm2extents is not a bitwise copy of the original RPM,
> > as such signature/digest verifications on anything that includes the
> > payload will not work.

> A better approach would be to change the RPM format to provide a digest
> that can be verified in a streaming manner.
>

The problem is not about verifying in a streaming manner, this is what is
currently done. The problem is that the final transcoded rpm (after it has
been verified) will not be able to pass `rpmkeys` anymore because it is not
a bit wise copy of the original rpm anymore.
I think what you mean by changing the rpm format is so that we can verify
the rpm in chunks while streaming it vs in its integrity, this would be a
different discussion and does not change the fact that once we have the
resulting transcoded rpm, it will not pass the original signature
verification because that rpm has now changed (for the purpose of
illustration let say this is equivalent of having the original rpm with its
payload compressed and storing it with its payload uncompressed, content of
files that would eventually make it to the disk are the same, but the rpm
is not a bit wise copy of the one we downloaded from the repo).

**Reply by DemiMarie** [reply]
> I think what you mean by changing the rpm format is so that we can verify
the rpm in chunks while streaming it

That is exactly what I mean.

**Comment by chantra** [comment]
@pmatilai do you have any feedback on this?

**Comment by chantra** [comment]
@pmatilai, checking back if you have any feedback on this? Is there anything you need more details on?

**Comment by pmatilai** [comment]
> Another new concept that needs to be introduced to the RPM plugin API is the ability for a plugin hook to stop the plugin chain from being executed. Currently, every plugin that registers to a specific hook is called as part of the plugin chain, regardless of whatever action may have been performed by a previous plugin. In cases such as installing file or creating the rpmfi iterator, as soon as a plugin has performed that action, the chain must be interrupted. In the current code, this is done by the introduction of the RPMRC_PLUGIN_CONTENTS return code.

I already said this in the original PR but maybe it got lost in the noise: this needs a different kind of API, interrupted chains is not a supportable concept. Something that allows the plugin to register as the owner of this payload format and the handler(s) called at appropriate times, with appropriate data. Such as the package file descriptor, and such as to actually install the file by reflinking it - AFAICS fsmUnpack(), or actually rpmfiArchiveReadToFilePsm() is the precise thing you'd want to replace with your own: you get a writable fd and you get to populate it with content by whatever means, while letting rpm handle all the other logic. That's about the extent of control I'm willing to consider handling to plugins.

So... what it all really comes down to is ability to define a custom archive type (see rpmfiNewArchiveReader()), and register them so that rpm knows to call it from fsmIter() in place of the internal default.

**Reply by pmatilai** [reply]
The reason this feels like trying to put a square peg through a round hole is that it's exactly what it is. The current plugin API was never intended to handle such a use-case, and so eg that RPMRC_PLUGIN_CONTENT is just trying to file off the corners a bit to make it somehow squeeze through :sweat_smile: 

This probably calls for a new plugin type to be introduced, one that specifically handles content only.

**Reply by chantra** [reply]
Thanks @pmatilai , this is really useful feedback.

I think I see how that could work. I will try to move to this model and report back as I hit roadblocks :)

**Comment by ddiss** [comment]
Thanks for working on this - the new CoW extent based approach looks exciting. I have a few comments / questions on the initial proposal...

> Files are converted (“transcoded”) locally during download using /usr/bin/rpm2extents (part of rpm codebase). The format is not intended to be “portable” - i.e. copying the files from the cache is not supported.
>
> Regular RPMs use a compressed .cpio based payload. In contrast, extent based RPMs contain uncompressed data aligned to the fundamental page size of the architecture, e.g. 4KiB on x86_64. This alignment is required for FICLONERANGE to work. Only files are represented in the payload, other directory entries like symlinks, device nodes etc are constructed entirely from rpm header information. Files are referenced by their digest, so identical files are de-duplicated.

I just wanted to highlight that, although not optimal, cpio archives can still be used alongside reflinks if the `newc` spec is *bent* a little to provide file data-segment alignment. The kernel initramfs archive format is quite static, so for [dracut-cpio](https://github.com/dracutdevs/dracut/pull/1531) (`dracut --enhanced-cpio`) reflinks I added functionality to zero-pad file names so that subsequent file data segments are block aligned. kernel and GNU cpio extractors handle this fine, with the caveat that padding can't exceed `PATH_MAX`.

On a separate note, Btrfs recently gained support for writing compressed extents directly to disk via the new `BTRFS_IOC_ENCODED_WRITE` ioctl. IIUC, Fedora also recently changed to using zstd by default for Btrfs root and rpms. Have you considered using `BTRFS_IOC_ENCODED_WRITE` functionality instead of performing decompression during download?

**Reply by malmond77** [reply]
The idea of aligning cpio metadata is very interesting. I can see how it'd help initramfs building speed tremendously.

As I understand it, RPM is pretty different: the main difference is that we're trying (fairly hard) not to change the normal format of rpm as found on mirrors for now. There are some very interesting ideas on how to change the upstream format, but in doing so, we'd render all existing servers unable to read the format.

If we could tolerate the breakage: I'd love to experiment with `BTRFS_IOC_ENCODED_WRITE` which would reduce writes down and eliminate explicit decompression. For clients or filesystems without CoW support: RPM could decompress and write the normal file. I was hoping encoded writes would eliminate the complex path with curl -> librepo -> rpm2extents. I'm not sure you could get data from the network and write encoded data to disk in one pass like we're doing now. Do you have any ideas on how to resolve that challenge?

Adding cpio metadata, along with a "null" compression type could help eliminate the change in `fsm.c` on how the payload is iterated. Note that `rpm2extents` does not (and cannot) touch headers without invalidating signatures, so the change in compression type is inferred and handled in the plugin.

Lastly, there's another optimization that would be lost in adopting cpio formatting: content de-duplication. I'm not sure how important this is tho in the big picture, so it might be a worthwhile tradeoff.

Thanks for the feedback! Matthew.

**Reply by ddiss** [reply]
> The idea of aligning cpio metadata is very interesting. I can see how it'd help initramfs building speed tremendously.
> 
> As I understand it, RPM is pretty different: the main difference is that we're trying (fairly hard) not to change the normal format of rpm as found on mirrors for now. There are some very interesting ideas on how to change the upstream format, but in doing so, we'd render all existing servers unable to read the format.

To clarify, aligning cpio data segments for *newly built* rpms shouldn't necessarily require any change in format. They'd continue to function the same as earlier cpio payload rpms, be it with some extra zero-padding.

> If we could tolerate the breakage: I'd love to experiment with `BTRFS_IOC_ENCODED_WRITE` which would reduce writes down and eliminate explicit decompression. For clients or filesystems without CoW support: RPM could decompress and write the normal file. I was hoping encoded writes would eliminate the complex path with curl -> librepo -> rpm2extents. I'm not sure you could get data from the network and write encoded data to disk in one pass like we're doing now. Do you have any ideas on how to resolve that challenge?

I'm not too familiar with the rpm on-disk format, but I'd hoped that `BTRFS_IOC_ENCODED_WRITE` could be used without a change to the format, by having the rpm header parsed during download to determine whether the compressed payload could be written as-is. With a cpio payload it'd then be a matter of copy_file_range()ing the (optimally aligned) compressed file data segments into the destination during installation.

`BTRFS_IOC_ENCODED_WRITE` appears very restrictive at this stage though:
- it requires `CAP_SYS_ADMIN`, so probably isn't a viable option for containers, etc.
- ioctl calls need to specify both unencoded and encoded offset+length, meaning that we'd still need to parse rpm payload compression metadata
- the ioctl unencoded length can't exceed 128 KiB
- for zstd encoded I/Os, the ioctl data must represent "as a single zstd frame with the windowLog compression parameter set to no more than 17"
  - On openSUSE Tumbleweed I see some rpms currently using zstd compression level 19. IIUC, Fedora uses the same zstd level

> Adding cpio metadata, along with a "null" compression type could help eliminate the change in `fsm.c` on how the payload is iterated. Note that `rpm2extents` does not (and cannot) touch headers without invalidating signatures, so the change in compression type is inferred and handled in the plugin.
> 
> Lastly, there's another optimization that would be lost in adopting cpio formatting: content de-duplication. I'm not sure how important this is tho in the big picture, so it might be a worthwhile tradeoff.

Indeed. FWIW, I think your extent based approach offers a lot of worthwhile benefits, but just wanted to point out that something similarly CoW friendly (although less efficient) is possible without necessarily requiring invasive changes :-)

> Thanks for the feedback! Matthew.

Thanks for the response!

**Reply by ddiss** [reply]
I hacked together a proof of concept implementation which uses `BTRFS_IOC_ENCODED_WRITE` to write a zstd-compressed cpio payload directly to disk as-is, from a carefully aligned rpm. Compressed extents are then reflinked to the installation path.
presentation: https://www.youtube.com/live/qNGw8IKaqc0?si=vrLkk8Bi9Odfqm4Z&t=10325
rpm-rs based source: https://github.com/ddiss/rpm/tree/poc-btrfs-zstd-encoded-io-reflink-extract

The rpm payload format isn't modified, although there's a slight "bending" of the cpio/newc spec to use the filename field for padding. zstandard frames making up the compressed rpm payload explicitly carry both compressed and uncompressed lengths, to allow detection of filesystem-supported I/O sizes.

**Reply by DemiMarie** [reply]
Do you plan on doing streaming cryptographic verification?  See <https://github.com/rpm-software-management/rpm/discussions/2057#discussioncomment-5054477>.

**Reply by ddiss** [reply]
If what you mean is that the payload shouldn't be used before it's been checked against a digest in the cryptographically verified header, then yes. The question is what quantifies as "used".
The payload data needs to go somewhere while we're calculating the digest, in which case we could "stream" it to Btrfs in it's compressed, yet-to-be-verified form via `BTRFS_IOC_ENCODED_WRITE`. However, doing so would allow any concurrent (pre-verification) reads to effectively fuzz the btrfs/zstd decompression code path.
One protection mechanism might be to write out the payload in its regular (compressed) form before attempting to *switch* those regular extents to compressed extents on btrfs via some new `copy_file_range()` flag. The regular->compressed extent switch would only be attempted following successful payload verification.
I'm very much open to any other suggestions :-)

**Reply by ddiss** [reply]
One other thing I noticed is that the rpm header carries the digest of the compressed payload in addition to the uncompressed payload digest. Verification of the compressed payload alongside `BTRFS_IOC_ENCODED_WRITE` is relatively straightforward, but verifying the uncompressed payload would require decompression, which we're obviously trying to avoid with this optimization.
I think it's safe to only perform compressed payload verification - zstd frames can carry an optional xxHash64 hash, which should be suitable for correctness checking if desired.

**Reply by pmatilai** [reply]
The compressed and uncompressed digests are *alternatives*, both cannot be valid at the same time. Rpm calculates both during build, but at verification uses the one that matches (if any). This is to allow freely changing between uncompressed and compressed format of the same content, which "obviously" must be a legit case. So, as long as one matches, you don't need to worry about the other too much.

**Reply by ddiss** [reply]
Makes sense, thanks for the clarification.

**Reply by pmatilai** [reply]
> The rpm payload format isn't modified, although there's a slight "bending" of the cpio/newc spec to use the filename field for padding. zstandard frames making up the compressed rpm payload explicitly carry both compressed and uncompressed lengths, to allow detection of filesystem-supported I/O sizes.

Note that the cpio/newc format is on its way out, you don't want to design too much around that. In rpm v4 it's used for packages where all contained files are below 4GB, but v6 will always use the newer rpm-specific format which doesn't carry any file metadata in the payload: https://rpm-software-management.github.io/rpm/manual/format_v6.html#payload

**Reply by ddiss** [reply]
Is there any way to pad (for alignment) between file data segments in the v6 payload? IIUC, the rpm header carries file data payload offset and length information, so perhaps sparse / zero ranges in between might be possible?

**Reply by pmatilai** [reply]
There are no offsets stored, so I don't think there is.

**Reply by pmatilai** [reply]
Thing is, since the payload is typically compressed, offsets are useless because they'd just point to somewhere in the middle of a compression stream, you can't jump to it without reading the whole thing anyhow. If the files in the payload were individually compressed, it'd seem quite reasonable to have offsets stored. Of course that would likely loose something in the compression ratio.

**Reply by ddiss** [reply]
I think as long as the zstd frame boundaries align with individual file data segments then it should still work fine with `BTRFS_IOC_ENCODED_WRITE` and reflink, although we'd need to write out individual files immediately rather than staging the complete cpio archive.

**Reply by DemiMarie** [reply]
@ddiss: From my perspective, plugins should not deal with untrusted input, so “used” means “passed to the plugin”.  This means that data must be buffered in memory until it can be cryptographically verified.  This is not possible with the current digest format, but a new format can support very efficient streaming verification, requiring only _O(log N)_ or even _O(1)_ space.

A simple design is something like this:

```c
struct signed_block {
    uint16_t size; // size of this block, limited to 64KiB
                   // but must be at least 1KiB (except last block)
    char hash[64]; // Hash of the next block, or all zeros for EOF
    char payload[]; // size - 64 bytes
};
```

With this proposal, the payload digest only covers the first block in the payload.  Once the first block is hashed, the payload is released to the plugin, while the hash is stored by RPM to verify the next block.

Another option is to use per-block hashes stored in the RPM header.  This bloats the RPM header, but allows for hash generation and verification to happen in parallel.  Assuming a 1MiB block size, a 1TiB RPM (which is absolutely massive), a 512-bit hash, and the use of a byte array tag in the header (more efficient than a string array), this would require ((2<sup>40</sup> ÷ 2<sup>20</sup>) × (512 ÷ 8)) = 64MiB of space in the header, which is well within the limit of 256MiB.

Whichever solution is chosen, RPM should provide a way to calculate a hash of the package that includes everything except the payload, and which returns an error if there is no payload digest.  This can be included in metadata.

**Reply by ddiss** [reply]
@DemiMarie I wonder whether packaging fsverity (Merkle tree) metadata into the rpm header would be an option for performant block-based hashing. It'd also bloat the rpm header, but may have the benefit of allowing the metadata to be reused for post-installation integrity and authenticity protection (assuming the files remain read-only and reside on ext4/btrfs).
For my specific `BTRFS_IOC_ENCODED_WRITE` use case, I didn't intend to implement it as a plug-in, but as part of core. The implementation doesn't require modifications to the existing on-disk format, and the only FS specific behaviour is the actual encoded write ioctl calls.

**Reply by DemiMarie** [reply]
@ddiss fsverity would also be suitable.  If you go with this approach, I recommend also including the total length of the payload in the (signed) header, to avoid vulnerabilities where extra data somehow doesn’t get hashed.

**Reply by pmatilai** [reply]
v6 will have signed payload size information : https://github.com/rpm-software-management/rpm/pull/3017/commits/784bb9076d614da33d29123f5ef6236a57d38463

**Comment by rphibel** [comment]
Hello @pmatilai,

I work at Meta with @chantra and @malmond77 . I am implementing the new API for the RPM CoW plugin based on your comment. This is what I have implemented so far:

I defined 2 new fields of type `rpmPlugin` (with associated getters and setters) in `rpmte` structure:
- [`customArchiveReader`](https://github.com/rphibel/rpm/blob/8de078508954e005ba0c974389979070cfe9431d/lib/rpmte.c#L85)
- [`customFileInstaller`](https://github.com/rphibel/rpm/blob/8de078508954e005ba0c974389979070cfe9431d/lib/rpmte.c#L86)

When a plugin wants to register as an archive reader for a package, it sets the field `customArchiveReader` in the [`psm_pre` ](https://github.com/rphibel/rpm/blob/8de078508954e005ba0c974389979070cfe9431d/plugins/reflink.c#L165)stage.
Similarly, if it wants to register as a file installer it sets the `customFileInstaller` field.

Then in [`fsm.c`](https://github.com/rphibel/rpm/blob/8de078508954e005ba0c974389979070cfe9431d/lib/fsm.c#L849), if these fields are set, archive reading, file installation is deferred to the plugin.

Here is the code:
https://github.com/rphibel/rpm/compare/cow_denylist_rebased...cow_denylist_refactoring

Could you please let me know if this is what you had in mind?

**Comment by rphibel** [comment]
@pmatilai Hello, I was wondering if you had a chance to have a look at my proposal? Please let me know if this is unclear or if you need more details.

**Comment by pmatilai** [comment]
Okay so, trying to give a better idea of what I'm after:

The fact that RPMRC_PLUGIN_CONTENTS is still needed is a strong signal that it's not quite right. 
The way I'm imagining this is that you have a probe function for the content handler, that gets called exactly one for each and every package, early on in every single transaction,. For stock rpm content, it just returns pointers to internal functions (or perhaps a struct with these pointers). For other content, it returns the handler from the first plugin that claims the content type and if nothing handles it, it's obviously an error. This should happen at the stage where the transaction is populated, not during it. 

The fsm code for handling different types doesn't differ at all, it's just a function pointer that gets called. Ie there's no rpmpluginsCallFsmFileArchiveReader() or rpmpluginsCallFsmFilePrepare(), these are determined by the early probe. And so there are no rpmteSetCustomFileInstaller() or rpmteSetCustomArchiveReader() or any corresponding custom-call-functions either, it's always just a function pointer to whatever is handling the content.

**Reply by DemiMarie** [reply]
To add to this: Plugins should not get access to content that has not been verified yet.  That means creating a new method of cryptographic verification, one that allows streaming verification of the data.

**Reply by rphibel** [reply]
Thanks a lot for these comments. I think I have a better idea of what I need to do. I will work on a new implementation.

**Reply by pmatilai** [reply]
> To add to this: Plugins should not get access to content that has not been verified yet. 

Regardless of the technical requirements, this is a fine example of what a sane API looks like: it's not about just making X somehow possible, but building the right usage into the API. Often easier said than done of course.

**Comment by ddiss** [comment]
> Thing is, since the payload is typically compressed, offsets are useless because they'd just point to somewhere in the middle of a compression stream, you can't jump to it without reading the whole thing anyhow.

zstd frame headers can carry both compressed and uncompressed sizes, which makes it a little easier to seek around within the compressed payload.

> If the files in the payload were individually compressed, it'd seem quite reasonable to have offsets stored. Of course that would likely loose something in the compression ratio.

That's mostly what I did in my rpm-rs based poc - files aren't individually compressed, but the start and end of individual file data segments correspond to zstd frame boundaries, so individual files can be extracted or written via `BTRFS_IOC_ENCODED_WRITE`/reflinked without much effort.


> > The rpm payload format isn't modified, although there's a slight "bending" of the cpio/newc spec to use the filename field for padding. zstandard frames making up the compressed rpm payload explicitly carry both compressed and uncompressed lengths, to allow detection of filesystem-supported I/O sizes.
> 
> Note that the cpio/newc format is on its way out, you don't want to design too much around that. In rpm v4 it's used for packages where all contained files are below 4GB, but v6 will always use the newer rpm-specific format which doesn't carry any file metadata in the payload: https://rpm-software-management.github.io/rpm/manual/format_v6.html#payload

Returning to this, I suppose I could still pad `07070X` payloads relatively easily by injecting individual "padding" files into the archive. This would still be compatible with V4, although it might be nice to reserve a special *padding* path in the V6 spec which can be skipped over (where V4 would still extract / install it). Is the V6 draft still open for these kinds of proposals?
cc: @lnussel

**Comment by evan-goode** [comment]
>  The way I'm imagining this is that you have a probe function for the content handler, that gets called exactly one for each and every package, early on in every single transaction,. For stock rpm content, it just returns pointers to internal functions (or perhaps a struct with these pointers). For other content, it returns the handler from the first plugin that claims the content type and if nothing handles it, it's obviously an error. This should happen at the stage where the transaction is populated, not during it.

+1 to the idea of pluggable payload types and plugins claiming ownership of a payload type. I am imagining a variant of RPM CoW using composefs (https://github.com/rpm-software-management/rpm/discussions/4223) and I believe the API additions discussed so far would also be sufficient for a hypothetical rpm2composefs.

**Reply by DemiMarie** [reply]
I think the signature validation should be handled in the RPM core.  Otherwise, I agree.
